tools/httpbintest/src/lib.rs

```diff
@@ -0,0 +1,1304 @@
+        let i = self.issued_reqs.get(&stream_id).unwrap();
+        self.reqs[*i].resp_hdrs.extend_from_slice(headers);
+    }
+
+    /// Append data to the response body for an issued request.
+    pub fn add_response_body(
+        &mut self, stream_id: u64, data: &[u8], data_len: usize,
+    ) {
+        let i = self.issued_reqs.get(&stream_id).unwrap();
+        self.reqs[*i].resp_body.extend_from_slice(&data[..data_len]);
+    }
+
+    /// Execute the test assertion(s).
+    pub fn assert(&mut self) -> bool {
+        (self.assert)(&self.reqs)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::sync::Once;
+
+    static INIT: Once = Once::new();
+
+    fn run_test<T>(test: T) -> ()
+    where
+        T: FnOnce() -> () + std::panic::UnwindSafe,
+    {
+        INIT.call_once(|| {
+            env_logger::builder()
+                .default_format_timestamp_nanos(true)
+                .init()
+        });
+
+        let result = std::panic::catch_unwind(|| test());
+
+        assert!(result.is_ok())
+    }
+
+    fn endpoint() -> url::Url {
+        let endpoint = match std::env::var_os("HTTPBIN_ENDPOINT") {
+            Some(val) => val.into_string().unwrap(),
+
+            None => String::from("https://cloudflare-quic.com/b"),
+        };
+
+        url::Url::parse(&endpoint).unwrap()
+    }
+
+    fn jsonify(data: &[u8]) -> HttpBinResponseBody {
+        serde_json::from_slice(&data).unwrap()
+    }
+
+    fn do_test(
+        reqs: Vec<HttpBinReq>, assert: HttpBinTestAssert, concurrent: bool,
+    ) {
+        let mut test = HttpBinTest::new(endpoint(), reqs, assert);
+        test.set_concurrency(concurrent);
+        assert!(crate::runner::run(&mut test));
+    }
+
+    // Build a single request and expected response with status code
+    fn request_check_status(testpoint: &str, status: usize) -> Vec<HttpBinReq> {
+        let expect_hdrs = Some(vec![Header::new(":status", &status.to_string())]);
+
+        let mut url = endpoint();
+        let path = format!("{}/{}", url.path(), &testpoint);
+        url.set_path(&path);
+
+        vec![HttpBinReq::new("GET", &url, None, expect_hdrs)]
+    }
+
+    // Build a single request with a simple JSON body using the provided method
+    fn request_with_body(method: &str) -> Vec<HttpBinReq> {
+        let expect_hdrs = Some(vec![Header::new(":status", "200")]);
+
+        let mut url = endpoint();
+        let path = format!("{}/{}", url.path(), method.to_ascii_lowercase());
+        url.set_path(&path);
+
+        let req_body = serde_json::json!({"key1": "value1", "key2": "value2"});
+
+        let mut req = HttpBinReq::new(
+            &method.to_ascii_uppercase(),
+            &url,
+            Some(req_body.to_string().into_bytes()),
+            expect_hdrs,
+        );
+
+        req.hdrs
+            .push(Header::new("content-type", "application/json"));
+
+        vec![req]
+    }
+
+    fn assert_request_body(reqs: &[HttpBinReq]) -> bool {
```

Comment:
Why does this take a slice if it only checks a single request?

Reply:
This function needs to match the signature of `HttpBinTestAssert` which accepts a slice. Perhaps there is a way to rewrite code such that assert can signature can accept a slice or individual item...